test/import/FSHImporter.Profile.test.ts

```diff
@@ -1838,6 +1895,260 @@ describe('FSHImporter', () => {
+      });
+
+      it('should parse an insert rule with parameters that will use the same RuleSet more than once with different parameters each time', () => {
+        const input = `
+        Profile: ObservationProfile
+        Parent: Observation
+        * insert EntryRules (Recursive)
+        `;
+        const allDocs = importer.import([new RawFSH(input, 'Insert.fsh')]);
+        expect(loggerSpy.getAllMessages('error')).toHaveLength(0);
+        expect(allDocs).toHaveLength(1);
+        const doc = allDocs[0];
+        expect(doc.appliedRuleSets.size).toBe(4);
+
+        const firstEntryRules = doc.appliedRuleSets.get(
+          Immutable.List<string>(['EntryRules', 'Recursive'])
+        );
+        expect(firstEntryRules).toBeDefined();
+        expect(firstEntryRules.rules).toHaveLength(1);
+        assertInsertRule(firstEntryRules.rules[0], 'RecursiveRules', ['5']);
+
+        const recursiveRules = doc.appliedRuleSets.get(
+          Immutable.List<string>(['RecursiveRules', '5'])
+        );
+        expect(recursiveRules).toBeDefined();
+        expect(recursiveRules.rules).toHaveLength(2);
+        assertCardRule(recursiveRules.rules[0], 'interpretation', 0, '5');
+        assertInsertRule(recursiveRules.rules[1], 'EntryRules', ['BaseCase']);
+
+        const secondEntryRules = doc.appliedRuleSets.get(
+          Immutable.List<string>(['EntryRules', 'BaseCase'])
+        );
+        expect(secondEntryRules.rules).toHaveLength(1);
+        assertInsertRule(secondEntryRules.rules[0], 'BaseCaseRules', ['5']);
+
+        const baseCaseRules = doc.appliedRuleSets.get(
+          Immutable.List<string>(['BaseCaseRules', '5'])
+        );
+        expect(baseCaseRules).toBeDefined();
+        expect(baseCaseRules.rules).toHaveLength(1);
+        assertCardRule(baseCaseRules.rules[0], 'note', 0, '5');
+      });
+
+      it('should log an error when an insert rule has the wrong number of parameters', () => {
+        const input = `
+        Profile: ObservationProfile
+        Parent: Observation
+        * insert OneParamRuleSet (#final, "Final")
+        `;
+        importer.import([new RawFSH(input, 'Insert.fsh')]);
+        expect(loggerSpy.getLastMessage('error')).toMatch(
+          /Incorrect number of parameters applied to RuleSet/s
+        );
+        expect(loggerSpy.getLastMessage('error')).toMatch(/File: Insert\.fsh.*Line: 4/s);
+      });
+
+      it('should log an error when an insert rule with parameters refers to an undefined parameterized RuleSet', () => {
+        const input = `
+        Profile: ObservationProfile
+        Parent: Observation
+        * insert MysteriousRuleSet ("mystery")
+        `;
+        importer.import([new RawFSH(input, 'Insert.fsh')]);
+        expect(loggerSpy.getLastMessage('error')).toMatch(
+          /Could not find RuleSet with parameters named MysteriousRuleSet/s
+        );
+        expect(loggerSpy.getLastMessage('error')).toMatch(/File: Insert\.fsh.*Line: 4/s);
+      });
```

Comment:
I think it would be good to have negative tests for when there are parser errors when applying the ruleset.  For example:
```
CardRuleSet(path, min, max)
* {path} {minId}..{max}

Profile: MyObservation
Parent: Observation
* insert CardRuleSet(path with spaces, 1, *) // <-- what happens here?
* insert CardRuleSet(code, *, 1) // <-- oops, got min and max backwards and '*' isn't valid min
* insert CardRuleSet(code, 1, infinityAndBeyond) // <-- Buzz Lightyear
```

Reply:
Similarly, I think we want to test to ensure that if a line in a ruleset is valid in the parser, but invalid in the exporter (for example, a path that doesn't exist), that the reported error is appropriately tracked to the right locations (e.g., the insert rule where it was applied and the underlying ruleset rule that originated the error).

Reply:
I'm adding in a negative test as you described, but I think that the case of "a line in a ruleset is valid in the parser, but invalid in the exporter" falls under the scope of Export Parameterized RuleSet.

The tricky part I'm running into is how to usefully describe the error. For example, in the first insert rule you've described above, we get a good old-fashioned "no viable alternative" error from the parser. This happens when parsing the generated RuleSet, which means we've called `parseDoc` on a manually constructed block of text. The error is logged by a `FSHListener`, which happens before `parseDoc` returns. Once we're back in `parseGeneratedRuleSet`, it's too late to change what the error is. I think the best that could be managed here is if the listener kept an internal list of any calls to its `syntaxError` method, we could check that list once we're back in `parseGeneratedRuleSet` and emit an additional message along the lines of "at least one bad thing happened when I tried to insert a parameterized rule set on line N", which at least lets the user know what line the offending insert rule is on. And if the error that is logged while parsing the generated document is not handled by our `FSHListener`, then we'd have to do something else. I guess if we ask the logger before we start how many errors it has, and it has more of them after running `parseDoc`, we can assume something went wrong with the insert rule?

Reply:
I think I'm OK w/ starting w/ a reasonable solution (not too many hoops to jump through) and seeing what it looks like (i.e., if it's "good enough").  If making it perfect adds a lot of complication, then I think we need to weigh the tradeoffs.

Reply:
After reviewing the code and tests, I think we need to do _something_ special here.  My main concern at this point is that the error message indicates its source as the file w/ the insert rule, but the line number is relative to the in-memory doc going through the second parse.  I think it's OK to return incomplete information if we need to, but we shouldn't return incorrect information.

One potentially simple (?) solution might just be to change the doc filename that gets set when we do the 2nd parse.  If we can change the name to something that clearly indicates its the in-memory application of a specific ruleset rule, that might be useful.